packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -1095,11 +1116,12 @@ class TextPainter {
       return null;
     }
     // We cap the offset at the final index of plain text.
+    final int nextCodeUnitIdx = min(offset, plainTextLength - 1);
     final int nextCodeUnit = plainText.codeUnitAt(min(offset, plainTextLength - 1));
 
     // Check for multi-code-unit glyphs such as emojis or zero width joiner
     final bool needsSearch = _isHighSurrogate(nextCodeUnit) || _isLowSurrogate(nextCodeUnit) || nextCodeUnit == _zwjUtf16 || _isUnicodeDirectionality(nextCodeUnit);
-    int graphemeClusterLength = needsSearch ? 2 : 1;
+    int graphemeClusterLength = _getCharactersGraphemeClusterLengthAtIndex(plainText, nextCodeUnitIdx);
```

Comment:
This does not fully solve the problem, because what this code needs in a grapheme cluster is to find a range long enough that when passed to SkParagraph via the [Paragraph.getBoxesForRange] method, SkParagraph returns a nonempty list of text boxes instead of an empty list. And that does not always coincide with a grapheme cluster as identified by `characters`.

Specifically, there's an example in #122478: in the string "प्रा", `characters` identifies that as two grapheme clusters "प्" and "रा". But SkParagraph treats the whole string "प्रा" as a single grapheme cluster. If we call [Paragraph.getBoxesForRange] on the subrange that's just "प्" or just "रा", we get an empty list.

When that occurs, this code will still need to go through the search loop below in order to get a nonempty list `boxes`. And that means it will still be subject to the bugs in that loop that are fixed by #122480.